trunk/daemon/btgdaemon.man
.TH BTGDAEMON 1 "01 Nov 2006"
.SH NAME
btgdaemon \- daemon part of the btg project.
.SH SYNOPSIS
.B "btgdaemon" 
[-h] 
[-D] 
[-v] 
[--syntax]
[-c FILE|--config FILE]
[-n|--nodetach]
[-s FILE|--session-file FILE]
[-N]

.SH DESCRIPTION
.B "btgdaemon" 
acts as a backend which performs bittorrent operations for a number of clients.

.B "btgdaemon" 
accepts connections using a transport. Currently the following transports are implemented: TCP/IP, secure TCP/IP, XML-RPC and secure XML-RPC. The secure transports are using TLS, provided by GNU TLS.

The daemon 
.B "must" 
be started before using one of the clients.

.SH OPTIONS

.TP
.BI "-h|--help"
\- Print a short help.

.TP
.BI "-D"
\- Write extensive debug information.

.TP
.BI "-v" 
\- Be verbose.

.TP
.BI "--syntax"
\- Output syntax of the configuration file used to standard out.

.TP
.BI "-c FILE|--config FILE"
\- Configuration file, FILE, to use.

.TP
.BI "-n|--nodetach"
\- Do not detach from TTY.

.TP
.BI "-s FILE|--session-file FILE"
\- Use session file FILE.

.TP
.BI "-N"
\- Do not reload sessions.

.SH FILES
.TP
No config files are used by the daemon. 
.TP
~/.btg/daemon - the default IPC que filename used for accepting connections from clients.

.SH AUTHOR
Written by Michael Wojciechowski.

.SH "SEE ALSO"
btg(5), btg_config(5), btgcallback(5), btgcli(1), btgui(1), btgd(1)

.SH BUGS
No known bugs.


trunk/ChangeLog
BTG version 0.9.7
-----------------
The following bugs were fixed:

012055 File checking fails on torrents that contain directories.

BTG version 0.9.6
-----------------
The following bugs were fixed:

011971 Inaccuracy in command (Loading All Torrent Files from a Dir).
011902 Memory leak in external/ini.
011901 FBSD, memory leak when using boost/zlib.
011782 FBSD, btgncli uses ~60% of cpu.
011759 Update howto and configure script.
011719 WWWUI does not respond (hardcoded port).
011678 Add more XML-RPC unittests.
011674 Update XML-RPC in external.
011644 WebBTG not working with Internet Explorer.
011575 / 010064 daemon should check "seed" dir for download files.
011440 btgui, use a thread for updates.
011237 wwwui, unable to handle "+" in passwords.
011192 Document release procedure.
011187 DHT, add command line argument to enable/disable.
011186 Implement support for encryption.
011185 Implement support for partial downloads.
011184 Remove use of old libtorrent versions.
011183 Session saving, use xmlrpc.
011122 Update site/howto: Gentoo ebuild	documentation.
011097 btgpasswd, check for unique directories (request).
011095 Libtorrent 0.12, FreeBSD, asio and no SO_NOSIGPIPE support.
010506 Request: Clean all button (wwwui).
010112 Request: Total time on torrent (wwwui).
009299 Add support for global limit (wwwui).
	
BTG version 0.9.5
-----------------
The following bugs were fixed:

010989 Add command_last to makefile.
010906 Tarball includes unnecessary directories.
010889 Listening on a single port does not work.
010880 TCP/IP proto doesn't work.
010859 XMLRPC, add unittest.
010858 Typedefs, add float.
010830 Update the howto with autotools versions.
010828 Update howto: bash shall be used to run configure script.
010826 Man pages, wrong sections.
010814 external/xml-rpc, fix warnings.
010665 btg-config doesn't work on FC6.
010639 Doesn't build (event-callback) on FC6.
010588 Context_create should return the context_id.
010586 Remove debian directory.
010582 Can't complie on CentOS 4.4.

BTG version 0.9
---------------
The following bugs were fixed:

010473 btgui, use plotmm.
010437 Clean fails when torrent has only one file (libtorrent bug).
010424 btgui, add details display.
010423 Naming torrents (feature request).
010316 Add check before moving to seed directory.
010198 Daemon crash: same file referenced by distinct torrents.
010166 Daemon skips notifies from tracker.
010076 Display time remaining with days if > 24 hours.
010065 Add option to open torrent w/o immediate start (feature request).
010063 Update btg-config man page about moving config files.
009903 TCP close_wait state.

BTG version 0.8.5
-----------------
The following bugs were fixed:

001062 btg-config, use full path to dialog.
009975 Context::dataPresentInSeedDir, missing try/catch.
009961 Clients: visible session ID.
009953 btgpasswd, argument documentation conflicts.
009939 btgncli, fix clean menu entry.
009905 btggui, initial dialog.
009904 Update howto: global limits / lt cvs.
009897 Update howto: directories used.
009889 Use shared libraries.

BTG version 0.8
---------------
The following bugs were fixed:

009875 btgncli, fix username query.
009782 btgncli, help menus.
009750 btgncli, add file menu entry.
009742 btgncli, load colors from file.
009740 btgncli, add load keybindings from file.
009739 btgncli, quit dialog.
009738 btgncli, init screen.
009737 btgncli, screen resize.
009736 btgncli, documentation.
009575 Listen-to doesn't work correctly.
009524 Unittest fail on fbsd.
009375 HTTP transport is not stable.
009361 Add DHT support.
009321 Daemon man page missing --config argument.
009219 Wrong port reported to tracker.
009112 Doesn't compile with Libtorrent-.11RC1 release.
009092 Configure fails when no g++ installed.
008308 libtorrent 0.10 problem, allocate_ressources.
007472 Add system test.

BTG version 0.7
---------------
The following bugs were fixed:

009076 Periodic sessionsaving.
009030 Howto update (title changed).
009006 Daemon doesn't save sessions.
009005 Broken btg-config.
008897 Format of supported filters.
008894 Add timers.
008893 Unable to handle more than 1024 clients.
008892 SocketGroup is not fair.
008763 Fatal errors are not printed without debug.
008663 btgpasswd should create directories.

trunk/daemon/btgdaemon.cpp
/*
 * btg Copyright (C) 2005 Michael Wojciechowski.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/*
 * $Id$
 */

#include <bcore/project.h>

#include <bcore/addrport.h>

#include <bcore/command_factory.h>
#include <bcore/util.h>
#include <bcore/dbuf.h>
#include <bcore/crashlog.h>

#include <bcore/os/sleep.h>
#include <bcore/os/fileop.h>
#include <bcore/os/id.h>

#include <bcore/logger/logger.h>
// Loggers:
#include <bcore/logger/console_log.h>
#include <bcore/logger/file_log.h>
#include <bcore/logger/syslog_log.h>

#include "daemondata.h"

#include "dconfig.h"
#include "darg.h"
#include "filetrack.h"

#include "ipfilter/ipfilterif.h"
#include "ipfilter/levelone.h"
#include "ipfilter/emule.h"

#include <map>

extern "C"
{
#include "daemon.h"
#include <bcore/sigh.h>

#include <unistd.h>
#include <time.h>
} // extern "C"

#include <daemon/auth/auth.h>
#include <daemon/auth/passwd_auth.h>

#include "btgdaemon.h"
#include "daemonhandler.h"
#include <bcore/verbose.h>

#if BTG_OPTION_EVENTCALLBACK
#  include "callbackmgr.h"
#endif // BTG_OPTION_EVENTCALLBACK

#if BTG_OPTION_UPNP
#  include "upnp/upnpif.h"
#  if BTG_OPTION_USECYBERLINK
#    include "upnp/cyberlink.h"
#  endif // BTG_OPTION_USECYBERLINK
#endif // BTG_OPTION_UPNP

#include "modulelog.h"

using namespace btg::core;
using namespace btg::core::logger;
using namespace btg::daemon;

const std::string moduleName("dem");

int global_btg_run = 1;
int global_btg_hup = 0;

// Main file for the daemon.
int main(int argc, char* argv[])
{
#if BTG_DEBUG
   btg::core::crashLog::init();
#endif // BTG_DEBUG

   daemonData dd;

   time(&dd.daemonStartTime);

   // Parse command line arguments.
   dd.cla = new commandLineArgumentHandler(GPD->sDEFAULT_DAEMON_CONFIG());
   dd.cla->setup();

   if (!dd.cla->parse(argc, argv))
      {
         dd.destroy();
         return BTG_ERROR_EXIT;
      }

   // Before doing anything else, check if the user wants to get the
   // syntax of the configuration file.
   if (dd.cla->listConfigFileSyntax())
      {
         daemonConfiguration non_existing("non_existing");

         std::cout << non_existing.getSyntax() << std::endl;

         dd.destroy();
         return BTG_NORMAL_EXIT;
      }

   dd.configFile = GPD->sDEFAULT_DAEMON_CONFIG();

   if (dd.cla->configFileSet())
      {
         dd.configFile = dd.cla->configFile();
      }

   bool verboseFlag = dd.cla->beVerbose();

   VERBOSE_LOG(verboseFlag, "Verbose logging enabled.");

   MVERBOSE_LOG(moduleName, verboseFlag, "config: '" << dd.configFile << "'.");

   if (!checkFile("configuration file", dd.configFile))
      {
         dd.destroy();
         return BTG_ERROR_EXIT;
      }

   dd.config     = new daemonConfiguration(dd.configFile);

   if (!dd.config->read())
      {
         BTG_FATAL_ERROR("btgdaemon", "Could not open configuration file '" << dd.configFile << "'");

         dd.destroy();
         return BTG_ERROR_EXIT;
      }

   MVERBOSE_LOG(moduleName, verboseFlag, "config read from '" << dd.configFile << "'.");

   if (dd.config->getUseTorrentName())
     {
       VERBOSE_LOG(verboseFlag, "Sending status using torrent names.");
     }
   else
     {
       VERBOSE_LOG(verboseFlag, "Sending status using file names.");
     }

   if ((dd.config->getRunAsUser().size() > 0) &&
       (dd.config->getRunAsGroup().size() > 0) )
      {
         if (btg::core::os::ID::changePossible())
            {
               bool idChangeResult = btg::core::os::ID::changeUserAndGroup(
                                                                           dd.config->getRunAsUser(),
                                                                           dd.config->getRunAsGroup()
                                                                           );
               if (idChangeResult)
                  {
                     VERBOSE_LOG(verboseFlag, "Changed user/group to " <<
                                 dd.config->getRunAsUser() << "/" <<
                                 dd.config->getRunAsGroup() << ".");
                  }
            }
      }
   else
      {
         VERBOSE_LOG(verboseFlag, "Not changing user/group.");
      }

   if (btg::core::os::ID::changePossible())
      {
         // Make sure that the daemon will not run as root.
         VERBOSE_LOG(verboseFlag, "BTG will not run as root, sorry.");
         dd.destroy();
         return BTG_ERROR_EXIT;
      }

   // Determine if the directories specified in the config file are
   // valid.

#if BTG_OPTION_SAVESESSIONS
   dd.ss_enable        = dd.config->getSSEnable();
   dd.ss_timeout       = dd.config->getSSTimeout();
   dd.ss_filename      = dd.config->getSSFilename();
#endif // BTG_OPTION_SAVESESSIONS

#if BTG_OPTION_EVENTCALLBACK
   dd.callbackmgr      = new callbackManager();
#endif // BTG_OPTION_EVENTCALLBACK

   // Daemonize.
   if (!dd.cla->doNotDetach())
      {
         switch (do_daemonize())
            {
            case PID_OK:
               {
                  BTG_MNOTICE("parent exiting, child daemonizing");
                  return BTG_NORMAL_EXIT;
               }
               break;
            case PID_ERR:
               {
                  BTG_MNOTICE("daemonized");
               }
               break;
            case SETID_ERR:
               {
                  BTG_MERROR("setid failed. Abort");
                  return BTG_ERROR_EXIT;
               }
               break;
            case CHDIR_ERR:
               {
                  BTG_MERROR("chdir failed. Abort");
                  return BTG_ERROR_EXIT;
               }
               break;
            }
      }
   else
      {
         VERBOSE_LOG(verboseFlag, "Not detaching from TTY.");
      }

   // At this point, the daemon has deatached and opened either syslog
   // or a file for logging.

#if BTG_DEBUG
   // Set debug logging, if requested.
   if (dd.cla->doDebug())
      {
         logWrapper::getInstance()->setMinMessagePriority(logWrapper::PRIO_DEBUG);
      }
   else if (verboseFlag)
      {
         logWrapper::getInstance()->setMinMessagePriority(logWrapper::PRIO_VERBOSE);
      }
   else
      {
         // Default: only report errors.
         logWrapper::getInstance()->setMinMessagePriority(logWrapper::PRIO_ERROR);
      }
#else
   // No debug enabled, check for verbose flag.
   if (verboseFlag)
      {
         logWrapper::getInstance()->setMinMessagePriority(logWrapper::PRIO_VERBOSE);
      }
   else
      {
         // Default: only report errors.
         logWrapper::getInstance()->setMinMessagePriority(logWrapper::PRIO_ERROR);
      }
#endif // BTG_DEBUG

   switch (dd.config->getLog())
      {
      case logBuffer::STDOUT:
         {
            logWrapper* log = logWrapper::getInstance();
            log->setLogStream(new logStream(new consoleLogger()));
            break;
         }
      case logBuffer::LOGFILE:
         {
            std::string logfilename = dd.config->getLogFile();

            // Expand filename
            btg::core::os::fileOperation::resolvePath(logfilename);

            logWrapper* log = logWrapper::getInstance();
            fileLogger* fl  = new fileLogger(logfilename);

            if (fl->open())
               {
                  log->setLogStream(new logStream(fl));
                  BTG_MNOTICE("logfile opened");
               }
            else
               {
                  if (dd.cla->doNotDetach())
                     {
                        // Log to stdout.
                        logWrapper* log = logWrapper::getInstance();
                        log->setLogStream(new logStream(new consoleLogger()));
                        BTG_MNOTICE("opening logfile '" << logfilename << "' failed");
                     }
               }
            break;
         }
      case logBuffer::SYSLOG:
         {
            // Log to syslog.
            logWrapper* log = logWrapper::getInstance();
            log->setLogStream(new logStream(new syslogLogger()));
            break;
         }
      case logBuffer::UNDEF:
         {
            // No logging.
            break;
         }
      }

   // Initialize authorization.

   std::string auth_file = dd.config->getAuthFile();

   if (!checkFile("passwd file", auth_file))
      {
         dd.destroy();
         return BTG_ERROR_EXIT;
      }

   dd.auth = new btg::daemon::auth::passwordAuth(auth_file);

   if (!dd.auth->initialized())
      {
         BTG_FATAL_ERROR("btgdaemon", "Unable to load passwd file '" << auth_file << "'.");

         dd.destroy();
         return BTG_ERROR_EXIT;
      }

   // Determine which transport to use.
   switch (dd.config->getTransport())
      {
      case messageTransport::TCP:
         {
            addressPort ap = dd.config->getListenTo();
            if (!ap.valid())
               {
                  BTG_FATAL_ERROR("btgdaemon", "Invalid IPv4:port pair '" << ap.toString() << "'");
                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }

            dd.externalization = new btg::core::externalization::XMLRPC();

            dd.transport = new tcpTransport(dd.externalization,
                                            100 * 1024,
                                            FROM_SERVER,
                                            ap,
                                            1000 // 1s timeout
                                            );

            if (!dd.transport->isInitialized())
               {
                  BTG_FATAL_ERROR("btgdaemon", "Could not create a TCP/IP transport");

                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }
            BTG_MNOTICE("initialized a TCP/IP transport");
            break;
         }
      case messageTransport::STCP:
         {
            addressPort ap = dd.config->getListenTo();
            if (!ap.valid())
               {
                  BTG_FATAL_ERROR("btgdaemon", "Invalid IPv4:port pair '" << ap.toString() << "'");
                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }

            std::string CAfile   = dd.config->getCACert();
            std::string CertFile = dd.config->getCert();
            std::string keyFile  = dd.config->getPrivKey();

            if (!checkFile("certificate authority, PEM", CAfile))
               {
                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }

            if (!checkFile("certificate, PEM", CertFile))
               {
                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }

            if (!checkFile("private key, RSA", keyFile))
               {
                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }

            btg::core::os::gtlsGeneric::init();

            dd.externalization     = new btg::core::externalization::XMLRPC();
            dd.secureTransportData = new btg::core::os::gtlsGlobalServerData(CAfile, CertFile, keyFile);

            dd.transport           = new btg::core::secureTcpTransport(dd.externalization,
                                                                       dd.secureTransportData,
                                                                       100 * 1024,
                                                                       FROM_SERVER,
                                                                       ap,
                                                                       1000 // 1s timeout
                                                                       );

            if (!dd.transport->isInitialized())
               {
                  BTG_FATAL_ERROR("btgdaemon", "Could not create a STCP/IP transport");

                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }

            BTG_MNOTICE("initialized a STCP/IP transport");
            break;
         }
      case messageTransport::XMLRPC:
         {
            addressPort ap = dd.config->getListenTo();
            if (!ap.valid())
               {
                  BTG_FATAL_ERROR("btgdaemon", "Invalid ipv4:port pair '" << ap.toString() << "'");
                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }

            dd.externalization = new btg::core::externalization::XMLRPC();

            dd.transport = new httpTransport(dd.externalization,
                                             100 * 1024,
                                             FROM_SERVER,
                                             ap,
                                             1000 // 1s timeout
                                             );

            if (!dd.transport->isInitialized())
               {
                  BTG_FATAL_ERROR("btgdaemon", "Could not create a http transport");

                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }
            BTG_MNOTICE("initialized a XMLRPC transport");
            break;
         }
      case messageTransport::SXMLRPC:
         {
            addressPort ap = dd.config->getListenTo();
            if (!ap.valid())
               {
                  BTG_FATAL_ERROR("btgdaemon", "Invalid IPv4:port pair '" << ap.toString() << "'");
                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }

            std::string CAfile   = dd.config->getCACert();
            std::string CertFile = dd.config->getCert();
            std::string keyFile  = dd.config->getPrivKey();

            if (!checkFile("certificate authority, PEM", CAfile))
               {
                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }

            if (!checkFile("certificate, PEM", CertFile))
               {
                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }

            if (!checkFile("private key, RSA", keyFile))
               {
                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }

            btg::core::os::gtlsGeneric::init();

            dd.externalization     = new btg::core::externalization::XMLRPC();
            dd.secureTransportData = new btg::core::os::gtlsGlobalServerData(CAfile, CertFile, keyFile);

            dd.transport           = new btg::core::secureHttpTransport(dd.externalization,
                                                                        dd.secureTransportData,
                                                                        100 * 1024,
                                                                        FROM_SERVER,
                                                                        ap,
                                                                        1000 // 1s timeout
                                                                        );

            if (!dd.transport->isInitialized())
               {
                  BTG_FATAL_ERROR("btgdaemon", "Could not create a SXMLRPC transport");

                  dd.destroy();
                  return BTG_ERROR_EXIT;
               }

            BTG_MNOTICE("initialized a SXMLRPC transport");
            break;
         }
      case messageTransport::UNDEFINED:
         {
            BTG_FATAL_ERROR("btgdaemon", "Unable to recognize transport name, cannot continue.");
         }
      }

   if (!dd.transport)
      {
         BTG_FATAL_ERROR("btgdaemon", "No transport, cannot continue.");

         dd.destroy();
         return BTG_ERROR_EXIT;
      }

   if (verboseFlag)
      {
         std::string transportName;

         switch (dd.config->getTransport())
            {
            case messageTransport::TCP:
               transportName = "TCP/IP";
               break;
            case messageTransport::STCP:
               transportName = "secure TCP/IP";
               break;
            case messageTransport::XMLRPC:
               transportName = "XML-RPC";
               break;
            case messageTransport::SXMLRPC:
               transportName = "secure XML-RPC";
               break;
            case messageTransport::UNDEFINED:
               transportName = "undefined";
               break;
            }

         VERBOSE_LOG(verboseFlag, "Transport, " << transportName << ", created and initialized.");
      }

   // Create an object to keep track of added torrents.
   dd.filetrack = new fileTrack();

   // Create an object to keep track of connections
   dd.connHandler = new connectionHandler();

   // Load filters.
   dd.config->getFilterFilename(dd.filterFilename);
   dd.filterType = dd.config->getFilterType();

   if (btg::core::os::fileOperation::check(dd.filterFilename))
      {
         std::string filtername;

         switch (dd.filterType)
            {
            case IpFilterIf::IPF_LEVELONE:
               {
                  VERBOSE_LOG(verboseFlag, "Loading IP filter, level one.");
                  BTG_MNOTICE("creating filter: IPF_LEVELONE");
                  filtername = "level one";
                  dd.filter  = new levelOne(verboseFlag, dd.filterFilename);

                  break;
               }
            case IpFilterIf::IPF_EMULE:
               {
                  VERBOSE_LOG(verboseFlag, "Loading IP filter, emule.");
                  BTG_MNOTICE("creating filter: IPF_EMULE");
                  filtername = "emule";
                  dd.filter  = new Emule(verboseFlag, dd.filterFilename);
                  break;
               }
            case IpFilterIf::IPF_UNDEF:
               {
                  dd.filter = 0;
                  break;
               }
            }

         if (dd.filter)
            {
               if ((!dd.filter->initialized()) || (dd.filter->numberOfEntries() == 0))
                  {
                     BTG_MNOTICE("filter not initialized or empty. Deleted.");

                     delete dd.filter;
                     dd.filter = 0;
                  }
               else
                  {
                     BTG_MNOTICE("IP filter '" << filtername << "' created, contains " << dd.filter->numberOfEntries() << " entries.");
                     VERBOSE_LOG(verboseFlag,
                                 "IP filter '" << filtername << "' created, contains " << dd.filter->numberOfEntries() << " entries.");
                  }
            }
      }
   else
      {
         BTG_NOTICE("unable to load IP filter");
         VERBOSE_LOG(verboseFlag, "Unable to load IP filter.");
      }

   // Make sure that the provided torrent range is valid and that the
   // daemon gets at least two ports.

   // Get the port range used by libtorrent.
   dd.portRange = dd.config->getLTPortRange();

   if ((((dd.portRange.first < daemonData::MIN_LISTEN_PORT) || (dd.portRange.second < daemonData::MIN_LISTEN_PORT)) || 
        (dd.portRange.first > dd.portRange.second)))
      {
         BTG_FATAL_ERROR("btgdaemon", "Unable to use port range: " << dd.portRange.first << 
                         ":" << dd.portRange.second << ".");

         dd.destroy();
         return BTG_ERROR_EXIT;
      }

   if ((dd.portRange.second - dd.portRange.first) < 1)
      {
	BTG_FATAL_ERROR("btgdaemon", "Unable to use port range: " << dd.portRange.first << 
                         ":" << dd.portRange.second << ", at least two ports are required.");

         dd.destroy();
         return BTG_ERROR_EXIT;
      }

   // Init upnp.
#if BTG_OPTION_UPNP
   bool const upnpEnabled = dd.config->getUseUPnP();
   btg::daemon::upnp::upnpIf* upnpif = 0;
#  if BTG_OPTION_USECYBERLINK
   upnpif = new btg::daemon::upnp::cyberlinkUpnpIf(verboseFlag, dd.config->getLTListenTo());
#  endif // BTG_OPTION_USECYBERLINK

   if (upnpif != 0)
      {
         if (upnpEnabled)
            {
               VERBOSE_LOG(verboseFlag, "UPnP: attempting to forward ports.");
               switch (upnpif->open(dd.portRange))
                  {
                  case true:
                     VERBOSE_LOG(verboseFlag, "UPnP: port forwarding succeded.");
                     break;
                  case false:
                     VERBOSE_LOG(verboseFlag, "UPnP: port forwarding failed.");
                     break;
                  }
            }
         else
            {
               VERBOSE_LOG(verboseFlag, "Not using UPnP.");
            }
      }
#endif // BTG_OPTION_UPNP

   BTG_MNOTICE("init done");

   sigh_add(SIGINT, (void*)&global_signal_handler);
   sigh_add(SIGTERM, (void*)&global_signal_handler);
   sigh_add(SIGHUP, (void*)&global_signal_handler);

   daemonHandler* dh = new daemonHandler(&dd, verboseFlag);

   VERBOSE_LOG(verboseFlag, GPD->sDAEMON() << " version " << GPD->sVERSION() << ".");
   VERBOSE_LOG(verboseFlag, "Ready to accept clients.");

   while (global_btg_run == 1)
      {
         dh->readFromTransport();
         dh->checkTimeout();

         if(global_btg_hup == 1)
            {
               // SIGHUP received, reload user file
               VERBOSE_LOG(verboseFlag, "Reloaded users file.");
               dd.auth->reInit();

               global_btg_hup = 0;
            }
      }

   dh->shutdown();

   delete dh;
   dh = 0;

   BTG_MNOTICE("finished execution");
   VERBOSE_LOG(verboseFlag, "Finished execution.");

   if (dd.config->modified())
      {
         BTG_MNOTICE("saving config");
         dd.config->write();
      }

   dd.destroy();

#if BTG_OPTION_UPNP
   if (upnpif != 0)
      {
         if (upnpEnabled)
            {
               switch (upnpif->close())
                  {
                  case true:
                     VERBOSE_LOG(verboseFlag, "UPnP: port forwarding removed.");
                     break;
                  case false:
                     VERBOSE_LOG(verboseFlag, "UPnP: attempt to remove port forwarding failed.");
                     break;
                  }
            }
      }

   delete upnpif;
   upnpif = 0;
#endif // BTG_OPTION_UPNP

   projectDefaults::killInstance();

   logWrapper::killInstance();

   return BTG_NORMAL_EXIT;
}

bool checkFile(std::string const& _purpose, std::string & _filename)
{
   bool status = true;

   std::string errorString;
   if (!btg::core::os::fileOperation::check(_filename, errorString, false))
      {
         BTG_FATAL_ERROR("btgdaemon", "Filename (" << _purpose << ") '" << _filename << "', could not be used: " << errorString);
         status = false;
      }

   return status;
}
